File: Leetcode_Python/Blind75/Graphs/GraphValidTree.py
##Iterative DFS approach - time complexity: O(N + M), space complexity - O(N + M)
from typing import List


def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if not n: ##handling the edge case wehere there are no nodes in the tree
            return True ##a graph with no nodes is regarded as a tree
        
        adj_List = {i :[] for i in range(n)}  ##defining the adjacency list
        
        for n1, n2 in edges: ##adding neighbors to the adjancency list
            adj_List[n1].append(n2)
            adj_List[n2].append(n1)
             
        visit = set() ##taking a set to track visisted nodes
        def dfs(i, prev): ##dfs function
            if i in visit: ##checking for loops
                return False
            
            visit.add(i) ##adding the current node to visit set
            for j in adj_List[i]: ##checking for neighboring nodes of the current node in adjancency list
                if j == prev: ##ignoring the trivial loop in an undirected graph
                    continue
                if not dfs(j, i): ##returning False, if dfs is not true at any point
                    return False
            return True ##return true if dfs is true for the entire for loop
        
        return dfs(0, -1) and n == len(visit) ##checking if all the nodes are visited and the result of dfs is true i.e. there is no loop
            
        

File: Leetcode_Python/Blind75/Graphs/GraphValidTree.java
package Blind75.Graphs;
/*
 * You have a graph of n nodes labeled from 0 to n - 1. You are given an integer n and a list of edges 
 * where edges[i] = [ai, bi] indicates that there is an undirected edge between nodes ai and bi in the graph.
Return true if the edges of the given graph make up a valid tree, and false otherwise.
 */
public class GraphValidTree {
    //Approach 1: Iterative depth first search
    public boolean validTree(int n, int[][] edges) {
        List<List<Integer>> adjacencyList = new ArrayList<>();
        
        for(int i = 0; i < n; i++){
            adjacencyList.add(new ArrayList<>());
        }
        
        for(int[] edge: edges){
            adjacencyList.get(edge[0]).add(edge[1]);
            adjacencyList.get(edge[1]).add(edge[0]);
        }
        
        Map<Integer, Integer> parent = new HashMap<>();
        
        parent.put(0, -1);
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        
        while(!stack.isEmpty()){
            int node = stack.pop();
            
            for (int neighbor : adjacencyList.get(node)){
                if (parent.get(node) == neighbor){
                    continue;
                }
                
                if (parent.containsKey(neighbor)){
                    return false;
                }
                
                stack.push(neighbor);
                parent.put(neighbor, node);
            }
        }
        
        return parent.size() == n;
        
    }
    
}
